[BUG][go-server] Duplicate auto-generated operationIds produce uncompilable output (route/handler/service name mismatch)

<!-- 
Title: [BUG][go-server] Duplicate auto-generated operationIds produce uncompilable output (route/handler/service name mismatch)
-->

#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apitools.dev/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

When two paths share the same auto-generated `operationId` (e.g. `/foo` and `/foo/`, both `GET`, neither declaring an explicit `operationId`), the `go-server` generator produces output that **does not compile**.

The route map and the `*Router`/`*Servicer` interfaces (which use `{{operationId}}`) get the deduplicated name `FooGet_0`, but the controller handler method and the service stub method (which use `{{nickname}}`) get a **different** deduplicated name `FooGet_1`. The generated file then references names that don't exist:

```go
// api_default.go (routes)
"FooGet_0": Route{ "GET", "/foo/", c.FooGet_0 },   // c.FooGet_0 does not exist

// api_default.go (handlers)
func (c *DefaultAPIController) FooGet_1(...) { ... c.service.FooGet_1(...) } // c.service.FooGet_1 not declared

// api.go (interface) declares FooGet_0 (not FooGet_1)
```

This is a problem because the generated server cannot be built at all without hand-editing the generated files. `go-gin-server` has the same `{{nickname}}`/`{{operationId}}` split and is affected too.

##### openapi-generator version

- Reproducible on **7.9.0** (where the regression was introduced by #19715).
- Still reproducible on **master / 7.23.0** (latest at the time of writing). #19913 only addressed `[csharp][generichost]` and does **not** fix `go-server`/`go-gin-server`.

##### OpenAPI declaration file content or url

```yaml
openapi: 3.0.1
info:
  title: repro
  version: '1.0'
paths:
  /foo:
    get:
      responses:
        '200':
          description: ok
  /foo/:
    get:
      responses:
        '200':
          description: ok
```

##### Generation Details

CLI, `go-server` generator, default options (only `sourceFolder`/`outputAsLibrary` set so the service stub is also generated and the mismatch is visible there too):

```bash
openapi-generator generate \
  -i repro.yaml \
  -g go-server \
  -o out/ \
  --package-name server \
  --enable-post-process-file \
  -p sourceFolder=server,outputAsLibrary=true
```

##### Steps to reproduce

1. Save the spec above as `repro.yaml`.
2. Run the generation command from *Generation Details*.
3. Build the generated server:

```bash
cd out/server
go mod init repro
go get github.com/gorilla/mux
go build ./...
```

Actual output (build fails):

```
./api_default.go:59:6: c.FooGet_0 undefined (type *DefaultAPIController has no field or method FooGet_0)
./api_default.go:78:27: c.service.FooGet_1 undefined (type DefaultAPIServicer has no field or method FooGet_1)
./api_default_service.go:42:18: c.service.FooGet_1 undefined (type DefaultAPIServicer has no field or method FooGet_1)
```

Inspection of the generated files shows the divergence:

- Routes (`api_default.go`): `FooGet`, `FooGet_0` (route `FooGet_0` → `c.FooGet_0`)
- Handlers (`api_default.go`): `FooGet`, `FooGet_1` (handler `FooGet_1` calls `c.service.FooGet_1`)
- Interfaces (`api.go`): `FooGet`, `FooGet_0` (both `*Router` and `*Servicer`)
- Service stub (`api_default_service.go`): `FooGet`, `FooGet_1`

Expected output: `nickname` and `operationId` stay consistent after deduplication, so routes / interfaces / handlers / service stubs all use `FooGet` and `FooGet_0`. `go build ./...` succeeds.

##### Related issues/PRs

- #19715 — introduced the `operationId` uniqueness dedup (regression source).
- #19865 — same regression reported for Java / kotlin-spring (`_<number>` suffixes).
- #19913 — fix for `[csharp][generichost]` only; does **not** cover `go-server`/`go-gin-server`.
- #20496 — adjacent `go-server` "Required validation too strict" issue (different bug, same generator).

##### Suggest a fix

Root cause: there are **two independent deduplication passes** with **inconsistent counter conventions**, and the go-server templates use both fields in the same compilation unit.

1. `DefaultCodegen.addOperationToGroup` deduplicates `operationId` — **first duplicate gets `_0`**:
   ```java
   int counter = 0;
   for (CodegenOperation op : opList) {
       if (uniqueName.equals(op.operationId)) {
           uniqueName = co.operationId + "_" + counter; // _0
           counter++;
       }
   }
   co.operationId = uniqueName;
   ```

2. `DefaultGenerator.processOperations` (and `processWebhooks`) separately deduplicates `nickname` (gated by `addSuffixToDuplicateOperationNicknames`, which defaults to `true` in `DefaultCodegen`) — **first duplicate gets `_1`**:
   ```java
   int counter = 0;
   for (CodegenOperation op : ops) {
       String opId = op.nickname;
       if (opIds.contains(opId)) {
           counter++;                       // 1
           op.nickname += "_" + counter;    // _1
       }
       opIds.add(opId);
   }
   ```

3. The go-server templates split on the two fields: `controller-api.mustache` uses `{{operationId}}` for the route map (`c.{{operationId}}`) but `{{nickname}}` for the handler method and the `c.service.{{nickname}}` call; `api.mustache` uses `{{operationId}}` for both interfaces; `service.mustache` uses `{{nickname}}` for the stub method.

Since `operationId` is already unique per tag after `addOperationToGroup`, the second nickname pass is redundant; it should mirror `operationId` instead of re-deduplicating with a different counter. In `DefaultGenerator.processOperations` / `processWebhooks`:

```java
// check for nickname uniqueness
if (config.getAddSuffixToDuplicateOperationNicknames()) {
    // operationId is already deduplicated per tag in addOperationToGroup,
    // so mirror it into nickname. The previous separate _<counter> dedup used a
    // different counter convention (_1 vs _0) and diverged nickname from
    // operationId, which breaks templates (e.g. go-server) that use {{nickname}}
    // for handlers/stubs and {{operationId}} for routes/interfaces.
    for (CodegenOperation op : ops) {
        op.nickname = op.operationId;
    }
}
```

This fixes every generator whose templates use `{{nickname}}` (`go-server`, `go-gin-server`, …) in one place, and is a no-op for non-duplicate operationIds (where `nickname == operationId` already).

A narrower alternative is to switch the go-server / go-gin-server templates from `{{nickname}}` to `{{operationId}}` everywhere (`controller-api.mustache` + `service.mustache` for go-server; `controller-api.mustache` + `interface-api.mustache` for go-gin-server), but the `DefaultGenerator` fix above is more fundamental and matches the intent of `addSuffixToDuplicateOperationNicknames`.

--- a/modules/openapi-generator/src/main/resources/go-gin-server/controller-api.mustache
+++ b/modules/openapi-generator/src/main/resources/go-gin-server/controller-api.mustache
@@ -15,7 +15,7 @@ type {{classname}} struct {
 {{#isDeprecated}}
 // Deprecated
 {{/isDeprecated}}
-func (api *{{classname}}) {{nickname}}(c *gin.Context) {
+func (api *{{classname}}) {{operationId}}(c *gin.Context) {
 	// Your handler implementation
 	c.JSON(200, gin.H{"status": "OK"})
 }

--- a/modules/openapi-generator/src/main/resources/go-gin-server/interface-api.mustache
+++ b/modules/openapi-generator/src/main/resources/go-gin-server/interface-api.mustache
@@ -10,12 +10,12 @@ type {{classname}} interface {
 
 
     {{#operation}}
-    // {{nickname}} {{httpMethod}} {{{basePathWithoutHost}}}{{{path}}}{{#summary}}
+    // {{operationId}} {{httpMethod}} {{{basePathWithoutHost}}}{{{path}}}{{#summary}}
     // {{{.}}} {{/summary}}
     {{#isDeprecated}}
     // Deprecated
     {{/isDeprecated}}
-     {{nickname}}(c *gin.Context)
+     {{operationId}}(c *gin.Context)
 
     {{/operation}}
     {{/operations}}

--- a/modules/openapi-generator/src/main/resources/go-server/controller-api.mustache
+++ b/modules/openapi-generator/src/main/resources/go-server/controller-api.mustache
@@ -92,11 +92,11 @@ func (c *{{classname}}Controller) OrderedRoutes() []Route {
 
 {{#operations}}{{#operation}}
 
-// {{nickname}} - {{{summary}}}
+// {{operationId}} - {{{summary}}}
 {{#isDeprecated}}
 // Deprecated
 {{/isDeprecated}}
-func (c *{{classname}}Controller) {{nickname}}(w http.ResponseWriter, r *http.Request) {
+func (c *{{classname}}Controller) {{operationId}}(w http.ResponseWriter, r *http.Request) {
 	{{#hasFormParams}}
 	{{#isMultipart}}
 	if err := r.ParseMultipartForm(32 << 20); err != nil {
@@ -641,7 +641,7 @@ func (c *{{classname}}Controller) {{nickname}}(w http.ResponseWriter, r *http.Re
 	{{/isArray}}
 	{{/isBodyParam}}
 	{{/allParams}}
-	result, err := c.service.{{nickname}}(r.Context(){{#allParams}}, {{#isNullable}}{{#isBodyParam}}&{{/isBodyParam}}{{/isNullable}}{{paramName}}Param{{/allParams}})
+	result, err := c.service.{{operationId}}(r.Context(){{#allParams}}, {{#isNullable}}{{#isBodyParam}}&{{/isBodyParam}}{{/isNullable}}{{paramName}}Param{{/allParams}})
 	// If an error occurred, encode the error with the status code
 	if err != nil {
 		c.errorHandler(w, r, err, &result)

--- a/modules/openapi-generator/src/main/resources/go-server/service.mustache
+++ b/modules/openapi-generator/src/main/resources/go-server/service.mustache
@@ -19,12 +19,12 @@ func New{{classname}}Service() *{{classname}}Service {
 	return &{{classname}}Service{}
 }{{#operations}}{{#operation}}
 
-// {{nickname}} - {{summary}}
+// {{operationId}} - {{summary}}
 {{#isDeprecated}}
 // Deprecated
 {{/isDeprecated}}
-func (s *{{classname}}Service) {{nickname}}(ctx context.Context{{#allParams}}, {{paramName}} {{#isNullable}}*{{/isNullable}}{{dataType}}{{/allParams}}) (ImplResponse, error) {
-	// TODO - update {{nickname}} with the required logic for this service method.
+func (s *{{classname}}Service) {{operationId}}(ctx context.Context{{#allParams}}, {{paramName}} {{#isNullable}}*{{/isNullable}}{{dataType}}{{/allParams}}) (ImplResponse, error) {
+	// TODO - update {{operationId}} with the required logic for this service method.
 	// Add {{classFilename}}_service.go to the .openapi-generator-ignore to avoid overwriting this service implementation when updating open api generation.
 
 {{#responses}}
@@ -39,5 +39,5 @@ func (s *{{classname}}Service) {{nickname}}(ctx context.Context{{#allParams}}, {
 
 {{/dataType}}
 {{/responses}}
-	return Response(http.StatusNotImplemented, nil), errors.New("{{nickname}} method not implemented")
+	return Response(http.StatusNotImplemented, nil), errors.New("{{operationId}} method not implemented")
 }{{/operation}}{{/operations}}

--- a/modules/openapi-generator/src/test/java/org/openapitools/codegen/goginserver/GoGinServerCodegenTest.java
+++ b/modules/openapi-generator/src/test/java/org/openapitools/codegen/goginserver/GoGinServerCodegenTest.java
@@ -20,10 +20,12 @@ package org.openapitools.codegen.goginserver;
 import org.openapitools.codegen.DefaultGenerator;
 import org.openapitools.codegen.TestUtils;
 import org.openapitools.codegen.config.CodegenConfigurator;
+import org.testng.Assert;
 import org.testng.annotations.Test;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Paths;
 import java.util.List;
@@ -83,6 +85,42 @@ public class GoGinServerCodegenTest {
 
         TestUtils.assertFileContains(Paths.get(output + "/go/api_pet.go"),
                 "type PetAPI interface");
+    }
+
+    @Test
+    public void verifyDuplicateOperationIdsAreConsistent() throws IOException {
+        // Regression test for duplicate auto-generated operationIds (e.g. /foo and /foo/).
+        // go-gin-server's routers.mustache references handleFunctions.<Class>.{{operationId}}
+        // while interface-api.mustache/controller-api.mustache used to declare the method
+        // with {{nickname}}. nickname and operationId were deduplicated by two independent
+        // passes with inconsistent counter conventions (_1 vs _0), so routers referenced
+        // FooGet_0 while the handler/interface declared FooGet_1, producing uncompilable
+        // code. The fix makes those templates use {{operationId}} consistently.
+        File output = Files.createTempDirectory("test").toFile();
+        output.deleteOnExit();
+
+        final CodegenConfigurator configurator = createDefaultCodegenConfigurator(output)
+                .setInputSpec("src/test/resources/3_0/go-gin-server/duplicate-operation-ids.yaml");
+
+        DefaultGenerator generator = new DefaultGenerator();
+        List<File> files = generator.opts(configurator.toClientOptInput()).generate();
+        files.forEach(File::deleteOnExit);
+
+        // routers.go uses {{operationId}} to reference the handler method on the generated struct.
+        TestUtils.assertFileContains(Paths.get(output + "/go/routers.go"),
+                "handleFunctions.DefaultAPI.FooGet_0");
+        // controller-api.mustache now uses {{operationId}} for the handler method, matching
+        // the {{operationId}} reference in routers.go.
+        TestUtils.assertFileContains(Paths.get(output + "/go/api_default.go"),
+                "func (api *DefaultAPI) FooGet_0(");
+
+        // Negative assertion: the divergent _1 name must not appear anywhere.
+        String routers = new String(Files.readAllBytes(Paths.get(output + "/go/routers.go")), StandardCharsets.UTF_8);
+        String apiDefault = new String(Files.readAllBytes(Paths.get(output + "/go/api_default.go")), StandardCharsets.UTF_8);
+        Assert.assertFalse(routers.contains("FooGet_1"),
+                "nickname diverged from operationId (FooGet_1 present in routers.go)");
+        Assert.assertFalse(apiDefault.contains("FooGet_1"),
+                "nickname diverged from operationId (FooGet_1 present in api_default.go)");
     }
 
     private static CodegenConfigurator createDefaultCodegenConfigurator(File output) {

--- a/modules/openapi-generator/src/test/java/org/openapitools/codegen/goserver/GoServerCodegenTest.java
+++ b/modules/openapi-generator/src/test/java/org/openapitools/codegen/goserver/GoServerCodegenTest.java
@@ -25,6 +25,7 @@ import org.testng.annotations.Test;
 
 import java.io.File;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Paths;
 import java.util.List;
@@ -72,6 +73,58 @@ public class GoServerCodegenTest {
         // verify /getPath/{id} is second route
         Assert.assertEquals(Files.readAllLines(Paths.get(output + "/go/api_dev.go")).get(58), "\t\t\"GetById\": Route{");
 
+    }
+
+    @Test
+    public void verifyDuplicateOperationIdsAreConsistent() throws IOException {
+        // Regression test for duplicate auto-generated operationIds (e.g. /foo and /foo/).
+        // Previously nickname and operationId were deduplicated by two independent passes with
+        // inconsistent counter conventions (_1 vs _0), so the route map / interfaces used
+        // {{operationId}} (e.g. FooGet_0) while the controller handler / service stub used
+        // {{nickname}} (e.g. FooGet_1), producing uncompilable code.
+        File output = Files.createTempDirectory("test").toFile();
+        output.deleteOnExit();
+
+        final CodegenConfigurator configurator = createDefaultCodegenConfigurator(output)
+                .setInputSpec("src/test/resources/3_0/go-server/duplicate-operation-ids.yaml");
+
+        DefaultGenerator generator = new DefaultGenerator();
+        List<File> files = generator.opts(configurator.toClientOptInput()).generate();
+        files.forEach(File::deleteOnExit);
+
+        // The deduplicated operationId pair is FooGet and FooGet_0; routes, handlers and
+        // service stubs must all reference the same name (templates use {{operationId}}).
+        TestUtils.assertFileContains(Paths.get(output + "/go/api_default.go"),
+                "\"FooGet\": Route{");
+        TestUtils.assertFileContains(Paths.get(output + "/go/api_default.go"),
+                "\"FooGet_0\": Route{");
+
+        // Route handler reference (uses {{operationId}}) and handler method definition
+        // (uses {{nickname}}) must reference the same name.
+        TestUtils.assertFileContains(Paths.get(output + "/go/api_default.go"),
+                "c.FooGet_0,");
+        TestUtils.assertFileContains(Paths.get(output + "/go/api_default.go"),
+                "func (c *DefaultAPIController) FooGet_0(");
+        // Service call inside the handler must also use the deduplicated operationId.
+        TestUtils.assertFileContains(Paths.get(output + "/go/api_default.go"),
+                "c.service.FooGet_0(");
+
+        // The interface (api.go) and the service stub must declare FooGet_0 too.
+        TestUtils.assertFileContains(Paths.get(output + "/go/api.go"),
+                "FooGet_0(");
+        TestUtils.assertFileContains(Paths.get(output + "/go/api_default_service.go"),
+                "func (s *DefaultAPIService) FooGet_0(");
+
+        // Negative assertions: the divergent _1 name must not appear anywhere.
+        String apiDefault = new String(Files.readAllBytes(Paths.get(output + "/go/api_default.go")), StandardCharsets.UTF_8);
+        String api = new String(Files.readAllBytes(Paths.get(output + "/go/api.go")), StandardCharsets.UTF_8);
+        String apiService = new String(Files.readAllBytes(Paths.get(output + "/go/api_default_service.go")), StandardCharsets.UTF_8);
+        Assert.assertFalse(apiDefault.contains("FooGet_1"),
+                "nickname diverged from operationId (FooGet_1 present in api_default.go)");
+        Assert.assertFalse(api.contains("FooGet_1"),
+                "nickname diverged from operationId (FooGet_1 present in api.go)");
+        Assert.assertFalse(apiService.contains("FooGet_1"),
+                "nickname diverged from operationId (FooGet_1 present in api_default_service.go)");
     }
 
     private static CodegenConfigurator createDefaultCodegenConfigurator(File output) {

--- a/modules/openapi-generator/src/test/resources/3_0/go-gin-server/duplicate-operation-ids.yaml
+++ b/modules/openapi-generator/src/test/resources/3_0/go-gin-server/duplicate-operation-ids.yaml
@@ -1,0 +1,21 @@
+openapi: 3.0.1
+
+info:
+  version: 1.0.0
+  title: Duplicate auto-generated operationIds
+
+paths:
+  /foo:
+    get:
+      tags:
+        - default
+      responses:
+        '200':
+          description: ok
+  /foo/:
+    get:
+      tags:
+        - default
+      responses:
+        '200':
+          description: ok

--- a/modules/openapi-generator/src/test/resources/3_0/go-server/duplicate-operation-ids.yaml
+++ b/modules/openapi-generator/src/test/resources/3_0/go-server/duplicate-operation-ids.yaml
@@ -1,0 +1,21 @@
+openapi: 3.0.1
+
+info:
+  version: 1.0.0
+  title: Duplicate auto-generated operationIds
+
+paths:
+  /foo:
+    get:
+      tags:
+        - default
+      responses:
+        '200':
+          description: ok
+  /foo/:
+    get:
+      tags:
+        - default
+      responses:
+        '200':
+          description: ok

